src/coreclr/jit/lsraxarch.cpp

```diff
@@ -2347,9 +2349,9 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree)
                     srcCount += BuildOperandUses(op2);
                     srcCount += BuildDelayFreeUses(op3, op1);
                 }
-                else if (op1->isContained())
+                else if (flag == 3)
                 {
-                    // 231 form: op3 = (op2 * op3) + [op1]
+                    // 231 form: op1 = (op2 * op3) + [op1]
```

Comment:
Put another way to help with clarity. Say you have:
```csharp
c = Fma.MultiplyAdd(a, b, c);
```

In this example, the operation done is logically `c = (a * b) + c`. The addend operand is always going to be `c` (`op3`) while the multiplication operands will always be `op1` or `op2` (and are commutative).

Since `op3` is the addend and also the destination, we want to pick `231` form: `xmm1 = (xmm2 * xmm3/m128) + xmm1`. Read another way, this is one of:
* `op3Reg = (op1Reg * op2Reg/m128) + op3Reg`: `vfmadd231ps op3Reg, op1Reg, op2Reg/m128`
* `op3Reg = (op2Reg * op1Reg/m128) + op3Reg`: `vfmadd231ps op3Reg, op2Reg, op1Reg/m128`

That is, the `operand #` does not necessarily match the register order displayed by the disassembly. Since the multiplication part is commutative, we can pick `op1` or `op2` to be the contained operand and therefore come from memory, rather than a register. Ideally only one of the operands is coming from memory and the choice is obvious, otherwise we just have to pick one.

The same logic applies to the
* `132` encoding, which is picked when:
  * `a` is the destination and `b` is coming from memory
  * `b` is the destination and `a` is coming from memory
* `213` encoding, which is picked when:
  * `a` is the destination and `c` is coming from memory
  * `b` is the destination and `c` is coming from memory

If none of the operands (`a`, `b`, or `c`) are the destination, then we only need to pick based on which operand needs to come from memory.
If none of the operands are the destination or come from memory, then we can just pick one form to use as the "default".

Reply:
This is a fantastic explanation that clears lot of confusion.

Reply:
@weilinwa, I would rewrite the comments to something like this (just that in future we remember how this works.:

```c++
// 231 form: xmm1 = (xmm2 * xmm3/mem) + xmm1
// One of the following operation:
// op3 = (op1 * [op2]) + op3
// op3 = (op2 * [op1]) + op3

// 132 form: xmm1 = (xmm1 * xmm3/mem) + xmm2
// One of the following operation:
// op1 = (op1 * [op2]) + op3
// op2 = (op2 * [op1]) + op3

// 213 form: xmm1 = (xmm1 * xmm2) + xmm3/mem
// op1 = (op1 * op2) + [op3]
```


For `132` comment, in the code that you have, you need to specify that under `(overwrittenOpNum == 1)`, you are just taking care of 1st case and 2nd case will be handled in last else.

Reply:
As you see, we need to make sure that we capture each scenario (which operand is in memory) and it might not be possible to do so by just checking for `overwritteOpNum`. We need to take into account which operand is contained.

```c++

if (overwrittenOpNum == 1)
{
    // 132 form: xmm1 = (xmm1 * xmm3/mem) + xmm2
    // op1 = (op1 * [op2]) + op3

    tgtPrefUse = BuildUse(op1);
    assert(op2->isContained());
    assert(!op3->isContained());

    srcCount += 1;
    srcCount += BuildOperandUses(op2);
    srcCount += BuildDelayFreeUses(op3, op1);
}
else if (overwrittenOpNum == 3)
{
    // 231 form: xmm1 = (xmm2 * xmm3/mem) + xmm1
    // One of the following operation:
    // op3 = (op1 * [op2]) + op3
    // op3 = (op2 * [op1]) + op3

    tgtPrefUse = BuildUse(op3);
  
    if (op2->IsContained())
    {
      srcCount += BuildOperandUses(op2);
      srcCount += BuildDelayFreeUses(op1, op3);
    }
    else if (op1->IsContained())
    {
      srcCount += BuildOperandUses(op1);
      srcCount += BuildDelayFreeUses(op2, op3);
    }
    else
    {
        assert(!"Both op1 and op2 are not contained");
    }

    srcCount += 1;
}
else
{
    assert(overwrittenOpNum == 2 || overwrittenOpNum == 0);

    if (op3->isContained())
    {
        // 213 form: xmm1 = (xmm1 * xmm2) + xmm3/mem
        // op1 = (op1 * op2) + [op3]
        
        tgtPrefUse = BuildUse(op1);
        srcCount += 1;
        
        srcCount += BuildOperandUses(op3);
        //TODO: Need to verify this
        if (copiesUpperBits)
        {
            srcCount += BuildDelayFreeUses(op2, op1);
        }
        else
        {
            tgtPrefUse2 = BuildUse(op2);
            srcCount += 1;
        }
    }
    else
    {
        // 132 form: xmm1 = (xmm1 * xmm3/mem) + xmm2
        // op2 = (op2 * [op1]) + op3
        
        assert(op1->isContained());
        
        tgtPrefUse = BuildUse(op2);
        srcCount += 1;
        
        srcCount += BuildOperandUses(op1);
        srcCount += BuildDelayFreeUses(op3, op2);
    }
}

```

@tannergooding - Just double check if I get it right.

Reply:
@kunalspathak, since we chose the format of fma in lowering and mark the corresponding op to contained, do we still need to handle different possibilities in lsra?

Reply:
In lowering, we marked the nodes to be contained or not, however in lsra, we need to build the ref positions for def/use and mark them delayfree as appropriate. We also want to make sure in lsra that we have correctly marked the operands as contained.

Reply:
Can I assume that when we mark op1 as contained in lowering for the case that ```overwrittenOpNum == 3```, in lsra, we will only go into the branch that ```op1->IsContained()``` when ```overwrittenOpNum == 3```? If this assumption is correct, we don't need to have the branch that ```if (op2->IsContained())```.

Reply:
From my understanding you might have to check both `op1` and `op2` as `IsContainableHWIntrinsicOp()` and depending on that set the `IsContained()` appropriately. Because for `231 form`, any of `op1` or `op2` can be in memory. 
@tannergooding - is that right?

Reply:
Correct. For `(a * b) + c`, `a` and `b` are commutative.

This means that `231` or `132` are both valid options, depending on if `c` or the other multiplicand are the destination.